File: ggcommons/metrics/metric.py
class Metric:
    # CloudWatch allows at most 10 dimensions per metric.
    MAX_DIMENSIONS = 10

    def __init__(
        self,
        thing_name: str,
        component_name: str,
        name: str,
        namespace: str = None,
        measures: dict = None,
        dimensions: dict = None,
    ):
        # Copy the incoming dicts so we never mutate a caller-owned collection when
        # injecting the default dimensions below.
        self.name = name
        self.namespace = namespace
        self.measures = dict(measures) if measures else {}
        self.dimensions = dict(dimensions) if dimensions else {}

        # Add default dimensions
        self.dimensions["coreName"] = thing_name
        self.dimensions["category"] = name
        self.dimensions["component"] = component_name

    def add_measure(self, measure):
        self.measures[measure.name] = measure

    def add_dimension(self, name, value):
        # Enforce the CloudWatch 10-dimension cap on the Metric itself (not just in
        # the builder), so the limit holds regardless of how the metric is built.
        if name not in self.dimensions and len(self.dimensions) >= self.MAX_DIMENSIONS:
            raise ValueError(
                f"Cannot add dimension '{name}': a metric may have at most "
                f"{self.MAX_DIMENSIONS} dimensions"
            )
        self.dimensions[name] = value

    def dimensions_as_json(self, include_core_name=True) -> list:
        dimensions_list = [
            {"name": key, "value": value}
            for key, value in self.dimensions.items()
            if include_core_name or key != "coreName"
        ]
        return dimensions_list

    def dimensions_as_collection(self, large_fleet_workaround: bool = False):
        return [
            {
                "Name": key,
                "Value": "ALL"
                if large_fleet_workaround and key == "coreName"
                else value,
            }
            for key, value in self.dimensions.items()
        ]

    def get_name(self):
        return self.name

    def get_namespace(self):
        return self.namespace

    def get_measures(self):
        return self.measures

    def get_measure(self, name):
        return self.measures.get(name)

    def get_dimensions(self):
        return self.dimensions

File: src/main/java/com/aws/proserve/ggcommons/metrics/Metric.java
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
package com.aws.proserve.ggcommons.metrics;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import software.amazon.awssdk.services.cloudwatch.model.Dimension;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

/**
 * Represents a metric definition in the Greengrass metrics system.
 * Contains metric configuration including name, namespace, dimensions, and measures.
 */
public class Metric
{
    private final String name;
    private final String namespace;
    private final Map<String, Measure> measures;
    private final Map<String, String> dimensions;

    /**
     * Creates a new metric with complete configuration.
     *
     * @param name The name of the metric
     * @param namespace The metric namespace
     * @param measures The map of measure definitions
     * @param dimensions The map of dimension key-values
     */
    public Metric(String name, String namespace, Map<String, Measure> measures, Map<String, String> dimensions)
    {
        if (measures == null) {
            throw new IllegalArgumentException("Measures cannot be null. At least 1 measure must be defined for a metric.");
        }
        
        this.name = name;
        this.namespace = namespace;
        this.measures = measures;
        this.dimensions = dimensions != null ? dimensions : new HashMap<>();
        addDimension("category", name);
    }

    /**
     * Adds a measure to this metric.
     *
     * @param measure The measure to add
     */
    public void addMeasure(Measure measure)
    {
        measures.put(measure.getName(), measure);
    }

    public void addDimension(String name, String value)
    {
        if (!dimensions.containsKey(name) && dimensions.size() >= 10) {
            throw new IllegalArgumentException("Maximum of 10 dimensions allowed per metric");
        }
        dimensions.put(name, value);
    }

    public JsonArray dimensionsAsJson()
    {
        return dimensionsAsJson(true);
    }

    public JsonArray dimensionsAsJson(boolean includeCoreName)
    {
        JsonArray jsonArray = new JsonArray();
        for (Map.Entry<String, String> dimension : dimensions.entrySet())
        {
            if (dimension.getKey().equals("coreName"))
            {
                if (includeCoreName)
                {
                    JsonObject arrayElem = new JsonObject();
                    arrayElem.addProperty("name", dimension.getKey());
                    arrayElem.addProperty("value", dimension.getValue());
                    jsonArray.add(arrayElem);
                }
            }
            else
            {
                JsonObject arrayElem = new JsonObject();
                arrayElem.addProperty("name", dimension.getKey());
                arrayElem.addProperty("value", dimension.getValue());
                jsonArray.add(arrayElem);
            }
        }
        return jsonArray;
    }

    public Collection<Dimension> dimensionsAsCollection(boolean largeFleetWorkaround)
    {
        Collection<Dimension> retVal = new ArrayList<>();
        for (Map.Entry<String, String> entry : dimensions.entrySet())
        {
            Dimension dimension;
            if (entry.getKey().equals("coreName") && largeFleetWorkaround)
            {
                dimension = Dimension.builder()
                                     .name(entry.getKey())
                                     .value("ALL")
                                     .build();
            }
            else
            {
                dimension = Dimension.builder()
                                     .name(entry.getKey())
                                     .value(entry.getValue())
                                     .build();
            }
            retVal.add(dimension);
        }
        return retVal;
    }

    public Collection<Dimension> dimensionsAsCollection()
    {
        return dimensionsAsCollection(false);
    }

    /**
     * Gets the name of this metric.
     *
     * @return The metric name
     */
    public String getName()
    {
        return name;
    }

    public String getNamespace()
    {
        return namespace;
    }

    public Map<String, Measure> getMeasures()
    {
        return measures;
    }

    public Measure getMeasure(String name)
    {
        return measures.get(name);
    }

    public Map<String, String> getDimensions()
    {
        return dimensions;
    }
}